### PR TITLE
ci: improve cache effectiveness

### PR DESCRIPTION
**Pull request by lenaschoenburg**
- Only save caches on main, not on other branches
- Don't try to cache docker build stages that are never reused

### CODE REVIEW

File: .github/actions/build-docker/action.yml [hunk 1]
```diff
@@ -83,6 +83,7 @@ runs:
         push: ${{ inputs.push }}
         cache-from: type=gha,ignore-error=true
         cache-to: type=gha,mode=max,ignore-error=true
+        no-cache-filters: build,distball,dist
```
Comment (npepinpe) [inline .github/actions/build-docker/action.yml:86]: :+1: Wasn't aware of that

**Review by npepinpe — APPROVED** [review]
:+1:

**Comment by npepinpe** [comment]
bors merge

I suppose we need to backport this to stable branches?

**Comment by oleschoenburg** [comment]
bors merge

**Comment by oleschoenburg** [comment]
bors merge

**Pull request merged**